fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java

```diff
@@ -150,39 +152,57 @@ public Optional<QueryPredicate> toQueryPredicate(@Nonnull final CorrelationIdent
                 return tryBoxSelfAsConstantPredicate();
             }
         } else if (childrenCount == 2) {
-            // only binary comparison functions are commutative.
-            // one side of the relop can be correlated to the innermost alias and only to that one; the other one
-            // can be correlated (or not) to anything except the innermostAlias
-            final Value leftChild = it.next();
-            final Value rightChild = it.next();
-            final Set<CorrelationIdentifier> leftChildCorrelatedTo = leftChild.getCorrelatedTo();
-            final Set<CorrelationIdentifier> rightChildCorrelatedTo = rightChild.getCorrelatedTo();
-
-            if (leftChildCorrelatedTo.isEmpty() && rightChildCorrelatedTo.isEmpty()) {
-                return tryBoxSelfAsConstantPredicate();
-            }
-            if (rightChildCorrelatedTo.contains(innermostAlias) && !leftChildCorrelatedTo.contains(innermostAlias)) {
-                if (leftChildCorrelatedTo.isEmpty()) {
-                    final Object comparand = compileTimeEvalFn.apply(leftChild);
-                    return comparand == null
-                           ? Optional.empty()
-                           : Optional.of(new ValuePredicate(rightChild, new Comparisons.SimpleComparison(swapBinaryComparisonOperator(comparisonType), comparand)));
-                } else {
-                    return Optional.of(new ValuePredicate(rightChild, new Comparisons.ValueComparison(swapBinaryComparisonOperator(comparisonType), leftChild)));
+            final var leftChild = it.next();
+            final var rightChild = it.next();
+            final var leftChildCorrelatedTo = leftChild.getCorrelatedTo();
+            if (comparisonType == Comparisons.Type.IN) {
+                if (!(rightChild instanceof LiteralValue)) {
+                    return Optional.empty();
                 }
+                final var literalValue = ((LiteralValue<?>) rightChild).getLiteralValue();
+                if (!(literalValue instanceof List)) {
+                    return Optional.empty();
+                }
+                final var evaluatedValues = new ArrayList<>();
+                ((List<?>) literalValue).forEach(value -> {
+                    final var operand = compileTimeEvalFn.apply((Value) value);
+                    if (operand != null) {
+                        evaluatedValues.add(operand);
+                    }
```

Comment:
I don't think this is correct, I think if `null` (unknown) exists as an operand in an `in-list` then the result is `null`. i.e.

```
null in (1, 2, null) --> null
2 in (1,2,null) -> null
(null in (1,2,3, null)) is null --> true
```

Reply:
```
2 in (1,2,null) -> true
3 in (1,2,null) -> null
```

Reply:
I stand corrected, thanks!

Since the `IN` in a shorthand for disjunctions, and in Kleene logic `true or null` is `null`:

```
2 in (1, 2, null) <=> (2 = 1 OR 2 = 2 OR 2 = null)
                  <=> (false OR true OR null) (execution is free to shortcut execution at this point I think,...)
                  <=> (true OR null) 
                  <=> (true)
```